compiler/codegen/cg-type.cpp

```diff
@@ -187,6 +205,217 @@ void EnumType::codegenDef() {
+          // skip getABITypeAlign() in the common case of deferred fieldAlign
+          ! isDeferredAlignment(fieldAlign) &&
+          fieldAlign > (int)DL.getABITypeAlign(fieldType).value())
+      {
+        // We had relied on LLVM to align our fields up until this field,
+        // which LLVM may not align correctly. Restart with explicit padding.
+        params.clear(); aligns.clear(); GEPMap.clear();
+        buildStructFields(DL, ag, true, GEPMap, params, aligns,
+                          structSize, structAlignment);
+        return;
+      }
+
+      if (explicitPadding) {
+        fieldAlign = fieldSym->getABIAlignment(fieldType);
+        // add padding before the field if needed
+        if (currSize % fieldAlign > 0)
+          insertPadding(fieldAlign - currSize % fieldAlign);
+      }
+
+      if (explicitPadding)
+        addFieldWhenEP(field->cname, fieldType,
+                       DL.getTypeStoreSize(fieldType), fieldAlign);
+      else
+        addFieldWhenNotEP(field->cname, fieldType);
+
+    } // for_fields
+
+    if (params.empty()) {
+      // add dummyFieldToAvoidWarning which also
+      // stops us from doing 0-byte memory allocation
+      // (comes up with ioNewline and --no-local RA)
+      // TODO - don't ever allocate 0-byte structures
+      // example: DefaultComparator
+      if (explicitPadding)
+        addFieldWhenEP("_dummyFieldToAvoidWarning", padTy, 1, 1);
+      else
+        addFieldWhenNotEP("_dummyFieldToAvoidWarning",
+          llvm::Type::getInt8Ty(gContext->llvmContext())); // here padTy==0
+    }
+  } // if 'ag' is a class or record
+
+  // add padding at end if needed
+  if (explicitPadding) {
+    if (currSize % currAlignment > 0)
+      insertPadding(currAlignment - currSize % currAlignment);
+  }
+
+  structSize = currSize;
+  structAlignment = currAlignment;
+}
+
+// Now that we have an llvm::StructType, we can check its layout:
+// what we think vs. what LLVM thinks.
+static void checkStructFields(const llvm::DataLayout& DL,
+                              AggregateType* ag, llvm::StructType* stype,
+                              std::vector<llvm::Type*>& params,
+                              std::vector<unsigned>& aligns,
+                              uint64_t structSize, int structAlignment)
+{
+  // For the release we should hide these behind 'if (fVerify)'
+  // to reduce codegen time and to reduce the chance of a false alarm
+  // on an architecture that we have not tested.
+  if (structSize == 0) {
+    INT_ASSERT(aligns.size() == 0);                   // we did not populate it
+    INT_ASSERT(isDeferredAlignment(structAlignment)); // delegated to LLVM
+  } else {
+    INT_ASSERT(params.size() == aligns.size());  // we populated them together
+    const llvm::StructLayout* layout = DL.getStructLayout(stype);
+    // These are other ways to query the struct size - any of them useful?
+    //  DL.getTypeAllocSize(stype) and layout->getSizeInBytes()
```

Comment:
should this go behind `fVerify`? And maybe instead of asserts everywhere, user facing error messages?

Reply:
These are really asserts, meaning that a violation would mean a bug on my end, not user's.

I am not excited about hiding them behind fVerify because it means the checks will be skipped in the majority of testing. I am open to removing these checks altogether after a couple of weeks of testing, if this addresses your concerns.

Reply:
No I think keep them, whether or not they need to go behind `fVerify` probably depends on if they cause a major perf regression